# EDA - Explarotary Data Analysis Capstone Project

**In this project, we tried to analyze human resources data by performing data wrangling and EDA**  
By working on this project we will learn how to:  
* Perform statistical analysis on real world data
* Deal with missing data using pandas
* Perform one-hot encoding
* Perform scaling including normalization and standardization
* Define a function and apply it to a Pandas DataFrame column
* Perform Pandas operations and filtering
* Calculate and display correlation matrix and heatmaps

# TASK-1: IMPORT DATASET AND PERFORM BASIC STATISTICAL DATA ANALYSIS

In [219]:
# Import dependencies

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

In [220]:
# Lets force pandas to display all rows and columns. Otherwise we will anly see summarized output which is sometimes is not too useful

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [221]:
# we import the dataset and observe the first records

df = pd.read_csv('Human_Resources.csv')
df.head()

Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
0   41         5993.0       Yes      Travel_Rarely       1102   
1   49            NaN        No  Travel_Frequently        279   
2   37         2090.0       Yes      Travel_Rarely       1373   
3   33         2909.0        No  Travel_Frequently       1392   
4   27         3468.0        No      Travel_Rarely        591   

               Department  DistanceFromHome  Education EducationField  \
0                   Sales                 1          2  Life Sciences   
1  Research & Development                 8          1  Life Sciences   
2  Research & Development                 2          2          Other   
3  Research & Development                 3          4  Life Sciences   
4  Research & Development                 2          1        Medical   

   EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  HourlyRate  \
0              1             1.0                        2  Female          94   
1              1             2.0                        3    Male          61   
2              1             4.0                        4    Male          92   
3              1             5.0                        4  Female          56   
4              1             7.0                        1    Male          40   

   JobInvolvement  JobLevel                JobRole  JobSatisfaction  \
0               3         2        Sales Executive                4   
1               2         2     Research Scientist                2   
2               2         1  Laboratory Technician                3   
3               3         1     Research Scientist                3   
4               3         1  Laboratory Technician                2   

  MaritalStatus  MonthlyRate  NumCompaniesWorked Over18 OverTime  \
0        Single      19479.0                   8      Y      Yes   
1       Married      24907.0                   1      Y       No   
2        Single       2396.0                   6      Y      Yes   
3       Married      23159.0                   1      Y      Yes   
4       Married      16632.0                   9      Y       No   

   PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
0               11.0                3.0                         1   
1               23.0                4.0                         4   
2               15.0                3.0                         2   
3               11.0                3.0                         3   
4               12.0                3.0                         4   

   StandardHours  StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0             80                 0                  8                      0   
1             80                 1                 10                      3   
2             80                 0                  7                      3   
3             80                 0                  8                      3   
4             80                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [222]:
# Lets also have a look to the end of our dataset

df.tail()

Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
1465   36         2571.0        No  Travel_Frequently        884   
1466   39         9991.0        No      Travel_Rarely        613   
1467   27         6142.0        No      Travel_Rarely        155   
1468   49         5390.0        No  Travel_Frequently       1023   
1469   34         4404.0        No      Travel_Rarely        628   

                  Department  DistanceFromHome  Education EducationField  \
1465  Research & Development                23          2        Medical   
1466  Research & Development                 6          1        Medical   
1467  Research & Development                 4          3  Life Sciences   
1468                   Sales                 2          3        Medical   
1469  Research & Development                 8          3        Medical   

      EmployeeCount  EmployeeNumber  EnvironmentSatisfaction Gender  \
1465              1          2061.0                        3   Male   
1466              1          2062.0                        4   Male   
1467              1          2064.0                        2   Male   
1468              1          2065.0                        4   Male   
1469              1          2068.0                        2   Male   

      HourlyRate  JobInvolvement  JobLevel                    JobRole  \
1465          41               4         2      Laboratory Technician   
1466          42               2         3  Healthcare Representative   
1467          87               4         2     Manufacturing Director   
1468          63               2         2            Sales Executive   
1469          82               4         2      Laboratory Technician   

      JobSatisfaction MaritalStatus  MonthlyRate  NumCompaniesWorked Over18  \
1465                4       Married      12290.0                   4      Y   
1466                1       Married      21457.0                   4      Y   
1467                2       Married       5174.0                   1      Y   
1468                2       Married      13243.0                   2      Y   
1469                3       Married      10228.0                   2      Y   

     OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
1465       No               17.0                3.0                         3   
1466       No               15.0                3.0                         1   
1467      Yes               20.0                4.0                         2   
1468       No               14.0                3.0                         4   
1469       No               12.0                3.0                         1   

      StandardHours  StockOptionLevel  TotalWorkingYears  \
1465             80                 1                 17   
1466             80                 1                  9   
1467             80                 1                  6   
1468             80                 0                 17   
1469             80                 0                  6   

      TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
1465                      3                3               5   
1466                      5                3               7   
1467                      0                3               6   
1468                      3                2               9   
1469                      3                4               4   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
1465                   2                        0                     3  
1466                   7                        1                     7  
1467                   2                        0                     3  
1468                   6                        0                     8  
1469                   3                        1                     2

In [223]:
# lets see how many features and instances we have in our dataset

df.shape

(1470, 35)

**As we see above, we have 1470 instances with 35 features**

In [224]:
# lets see the summary statistics of our dataset's numericalfeatures
df.describe().T

count          mean          std     min      25%  \
Age                       1470.0     36.923810     9.135373    18.0    30.00   
MonthlyIncome             1467.0   6505.155419  4711.297846  1009.0  2911.00   
DailyRate                 1470.0    802.485714   403.509100   102.0   465.00   
DistanceFromHome          1470.0      9.192517     8.106864     1.0     2.00   
Education                 1470.0      2.912925     1.024165     1.0     2.00   
EmployeeCount             1470.0      1.000000     0.000000     1.0     1.00   
EmployeeNumber            1469.0   1025.556161   601.646166     1.0   492.00   
EnvironmentSatisfaction   1470.0      2.721769     1.093082     1.0     2.00   
HourlyRate                1470.0     65.891156    20.329428    30.0    48.00   
JobInvolvement            1470.0      2.729932     0.711561     1.0     2.00   
JobLevel                  1470.0      2.063946     1.106940     1.0     1.00   
JobSatisfaction           1470.0      2.728571     1.102846     1.0     2.00   
MonthlyRate               1468.0  14310.966621  7121.294449  2094.0  8043.75   
NumCompaniesWorked        1470.0      2.693197     2.498009     0.0     1.00   
PercentSalaryHike         1469.0     15.206263     3.659047    11.0    12.00   
PerformanceRating         1469.0      3.153165     0.360270     3.0     3.00   
RelationshipSatisfaction  1470.0      2.712245     1.081209     1.0     2.00   
StandardHours             1470.0     80.000000     0.000000    80.0    80.00   
StockOptionLevel          1470.0      0.793878     0.852077     0.0     0.00   
TotalWorkingYears         1470.0     11.279592     7.780782     0.0     6.00   
TrainingTimesLastYear     1470.0      2.799320     1.289271     0.0     2.00   
WorkLifeBalance           1470.0      2.761224     0.706476     1.0     2.00   
YearsAtCompany            1470.0      7.008163     6.126525     0.0     3.00   
YearsInCurrentRole        1470.0      4.229252     3.623137     0.0     2.00   
YearsSinceLastPromotion   1470.0      2.187755     3.222430     0.0     0.00   
YearsWithCurrManager      1470.0      4.123129     3.568136     0.0     2.00   

                              50%       75%      max  
Age                          36.0     43.00     60.0  
MonthlyIncome              4908.0   8378.00  19999.0  
DailyRate                   802.0   1157.00   1499.0  
DistanceFromHome              7.0     14.00     29.0  
Education                     3.0      4.00      5.0  
EmployeeCount                 1.0      1.00      1.0  
EmployeeNumber             1022.0   1556.00   2068.0  
EnvironmentSatisfaction       3.0      4.00      4.0  
HourlyRate                   66.0     83.75    100.0  
JobInvolvement                3.0      3.00      4.0  
JobLevel                      2.0      3.00      5.0  
JobSatisfaction               3.0      4.00      4.0  
MonthlyRate               14235.5  20463.25  26999.0  
NumCompaniesWorked            2.0      4.00      9.0  
PercentSalaryHike            14.0     18.00     25.0  
PerformanceRating             3.0      3.00      4.0  
RelationshipSatisfaction      3.0      4.00      4.0  
StandardHours                80.0     80.00     80.0  
StockOptionLevel              1.0      1.00      3.0  
TotalWorkingYears            10.0     15.00     40.0  
TrainingTimesLastYear         3.0      3.00      6.0  
WorkLifeBalance               3.0      3.00      4.0  
YearsAtCompany                5.0      9.00     40.0  
YearsInCurrentRole            3.0      7.00     18.0  
YearsSinceLastPromotion       1.0      3.00     15.0  
YearsWithCurrManager          3.0      7.00     17.0

**Goruldugu gibi bu datasetinde 1470 farkli calisanin, calisma hayatlariyla ilgiler bilgiler duzenlenmistir**

Kisa bir ozet yapmak gerekirse, calisanlarin yas ortalamasi 37 olup ortalama 6505 USD maas almaktadirlar. Ortalama, bir calisanin calisma suresi ise 11 yildir  

In [225]:
#sum prominent features mean values

mean_columns = ['Age', 'MonthlyIncome', 'TotalWorkingYears']
df.describe().T.loc[mean_columns, 'mean'].to_frame()

mean
Age                  36.923810
MonthlyIncome      6505.155419
TotalWorkingYears    11.279592

In [226]:
# lets also see some other metrics related to employee ages
# There is a subtle usage of ' and " here, lets notice this difference
# Remember to use " not ' to escape ' in the f string
# Also notice to write the code in 2 lines according to PEP rules, and printing out the results in 2 lines according to PEP rules

print(f"The average of employee age is {round(np.mean(df['Age']))}"
      f"\nwhere as the minimum and maximum ages are {np.min(df['Age'])} and {np.max(df['Age'])} consecutively")

The average of employee age is 37
where as the minimum and maximum ages are 18 and 60 consecutively


# TASK-2: DEALING WITH MISSING DATA

**One important note**  
Checking missing values before diving into the dataset is a crucial operation,  
however it's also important that all NaN values **may not be missing values**.  
Sometimes it might BE a value of a corresponding categorical value  
You should be aware of that

## Filtering ROWS having Nan or Null Values

Here, lets notice the usage of **axis** in any statement  
**df.isnull()** returns a dataframe with True and False values.  
**any(axis=1)** is filtering this dataframe --> Any ask each rows that is there any True value in a corresponding COLUMN (axis=1)  
**df.isnull().any()** default axis=0; ask the same question to each column whether it has any missing value at ANY rows   

In [227]:
print(f"There are {len(df[df.isnull().any(axis=1)])} rows with MISSING VALUES \n\n")

df[df.isnull().any(axis=1)]

#another version of this code
#df[df.isnull().sum(axis=1)>0]

There are 7 rows with MISSING VALUES 




Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
1     49            NaN        No  Travel_Frequently        279   
6     59            NaN        No      Travel_Rarely       1324   
7     30         2693.0        No      Travel_Rarely       1358   
17    22         2935.0        No         Non-Travel       1123   
23    21         1232.0        No      Travel_Rarely        391   
103   34         4809.0        No      Travel_Rarely        665   
947   52            NaN       Yes      Travel_Rarely       1030   

                 Department  DistanceFromHome  Education EducationField  \
1    Research & Development                 8          1  Life Sciences   
6    Research & Development                 3          3            NaN   
7                       NaN                24          1  Life Sciences   
17   Research & Development                16          2        Medical   
23   Research & Development                15          2  Life Sciences   
103  Research & Development                 6          4          Other   
947                   Sales                 5          3  Life Sciences   

     EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  \
1                1             2.0                        3    Male   
6                1             NaN                        3     NaN   
7                1            11.0                        4    Male   
17               1            22.0                        4    Male   
23               1            30.0                        3    Male   
103              1           138.0                        1  Female   
947              1          1319.0                        2    Male   

     HourlyRate  JobInvolvement  JobLevel                JobRole  \
1            61               2         2     Research Scientist   
6            81               4         1                    NaN   
7            67               3         1  Laboratory Technician   
17           96               4         1  Laboratory Technician   
23           96               3         1     Research Scientist   
103          41               3         2     Research Scientist   
947          64               3         3        Sales Executive   

     JobSatisfaction MaritalStatus  MonthlyRate  NumCompaniesWorked Over18  \
1                  2       Married      24907.0                   1      Y   
6                  1       Married       9964.0                   4      Y   
7                  3      Divorced      13335.0                   1      Y   
17                 4           NaN       7324.0                   1      Y   
23                 4        Single          NaN                   1      Y   
103                3        Single          NaN                   1      Y   
947                2        Single      21534.0                   9      Y   

    OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
1         No               23.0                4.0                         4   
6        Yes                NaN                NaN                         1   
7         No               22.0                4.0                         2   
17       Yes               13.0                3.0                         2   
23        No               14.0                3.0                         4   
103       No               14.0                3.0                         3   
947      Yes               19.0                3.0                         3   

     StandardHours  StockOptionLevel  TotalWorkingYears  \
1               80                 1                 10   
6               80                 3                 12   
7               80                 1                  1   
17              80                 2                  1   
23              80                 0                  0   
103             80                 0                 16   
947             80                 0                 10   

     TrainingT

At the results above it's obvious that **Row index 6** has many missing values. Lets also check which rows have how many missing values

In [228]:
df.isnull().sum(axis=1).iloc[df[df.isnull().sum(axis=1)>0].index]

1      1
6      7
7      1
17     1
23     1
103    1
947    1
dtype: int64

## Filtering COLUMNS having Nan or Null Values

In [229]:
nan_column_count = df.isnull().any().sum()
print(f"Totally {nan_column_count} columns have missing values ")

Totally 10 columns have missing values 


In [230]:
df.isnull().sum().nlargest(nan_column_count+1)

MonthlyIncome        3
MonthlyRate          2
Department           1
EducationField       1
EmployeeNumber       1
Gender               1
JobRole              1
MaritalStatus        1
PercentSalaryHike    1
PerformanceRating    1
Age                  0
dtype: int64

## Dropping the NaN values

In [231]:
df.dropna(axis=0, inplace=True)

In [232]:
df.shape

(1463, 35)

After dropping 7 columns that have any nan values, now we have **1463 instances**

In [233]:
# Lets double check whether we have any missing values

df.isnull().sum().nlargest(5)

Age               0
MonthlyIncome     0
Attrition         0
BusinessTravel    0
DailyRate         0
dtype: int64

## An Alternative Way for Missing Values: IMPUTATION

Since dropping any missing values are the easiest way to handling with missing values it may not be the smartest solution everytime.   
There are several ways  and one of them is imputation - on other words - filling them someway  
There are also several ways for imputation, for example:  
- Imputation with Mean/Median/Most Frequent Value
- Forward or Backward Fill
- Imputation with Machine Learning Models
- Using Statistical Methods
- Use of Domain Knowledge
etc..

In [267]:
# lets call back the DATA and missing value rows

df = pd.read_csv('Human_Resources.csv')
nan_column_count = df.isnull().any().sum()
print(f"Totally {nan_column_count} columns have missing values ")

Totally 10 columns have missing values 


In [268]:
df.isnull().sum().nlargest(nan_column_count)

MonthlyIncome        3
MonthlyRate          2
Department           1
EducationField       1
EmployeeNumber       1
Gender               1
JobRole              1
MaritalStatus        1
PercentSalaryHike    1
PerformanceRating    1
dtype: int64

Lets check **dataypes** of features having missing values. Then we will **impute** the missing values **accordingly**

In [269]:
missing_columns = df.isnull().sum().nlargest(nan_column_count).index

In [270]:
df[missing_columns].dtypes

MonthlyIncome        float64
MonthlyRate          float64
Department            object
EducationField        object
EmployeeNumber       float64
Gender                object
JobRole               object
MaritalStatus         object
PercentSalaryHike    float64
PerformanceRating    float64
dtype: object

We can simply fill float missing values by the mean of this feature **(except EmployeNumber - it should be an ordinal sequence)**  
and any categorical features (object) by the feature's mode

In [271]:
fill_with_mean_features = ['MonthlyIncome', 'MonthlyRate', 'PercentSalaryHike', 'PerformanceRating']
fill_with_mean_features

['MonthlyIncome', 'MonthlyRate', 'PercentSalaryHike', 'PerformanceRating']

In [273]:
df[fill_with_mean_features] = df[fill_with_mean_features].fillna(df.mean())

In [277]:
df[fill_with_mean_features].isnull().any().any()

False

In [278]:
# lets figure out how we can impute EmployeeNumber
df['EmployeeNumber'].head()

0    1.0
1    2.0
2    4.0
3    5.0
4    7.0
Name: EmployeeNumber, dtype: float64

In [279]:
df[df['EmployeeNumber'].isnull()==True]

Age  MonthlyIncome Attrition BusinessTravel  DailyRate  \
6   59    6505.155419        No  Travel_Rarely       1324   

               Department  DistanceFromHome  Education EducationField  \
6  Research & Development                 3          3            NaN   

   EmployeeCount  EmployeeNumber  EnvironmentSatisfaction Gender  HourlyRate  \
6              1             NaN                        3    NaN          81   

   JobInvolvement  JobLevel JobRole  JobSatisfaction MaritalStatus  \
6               4         1     NaN                1       Married   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
6       9964.0                   4      Y      Yes          15.206263   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
6           3.153165                         1             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
6                 3                 12                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
6                2               1                   0   

   YearsSinceLastPromotion  YearsWithCurrManager  
6                        0                     0

In [280]:
df['EmployeeNumber'].iloc[3:10]

3     5.0
4     7.0
5     8.0
6     NaN
7    11.0
8    12.0
9    13.0
Name: EmployeeNumber, dtype: float64

In [281]:
# So we can impute this missing value with exactly 9.0 value

df['EmployeeNumber'] = df['EmployeeNumber'].fillna(9.0)
df['EmployeeNumber'].isnull().any()

False

In [283]:
# lets see the filled value

df.iloc[6]['EmployeeNumber']

9.0

In [295]:
# Lets impute the object (categorical) features as well

fill_with_mode_features = df[missing_columns].select_dtypes(include=['object']).columns.tolist()
fill_with_mode_features

['Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus']

In [304]:
# Lets see the mode values of each object features with Nan values
df[fill_with_mode_features].mode()

Department EducationField Gender          JobRole MaritalStatus
0  Research & Development  Life Sciences   Male  Sales Executive       Married

In [307]:
# here, its a little bit different than filling wth mean. df.mode() returns a dataframe so we have to use iloc to filter the exact mode to fill

df[fill_with_mode_features] = df[fill_with_mode_features].fillna(df.mode().iloc[0])

In [309]:
# lets finally check again whether we have any remaining Nan values 

df.isnull().any().any()

False

# TASK-3: PERFORM ONE-HOT ENCODING [OPTIONAL]

# TASK-4: PERFORM SCALING (NORMALIZATION & STANDARDIZATION) [OPTIONAL]

# TASK-5: PANDAS OPERATIONS/FILTERING

In [310]:
df.head()

Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
0   41    5993.000000       Yes      Travel_Rarely       1102   
1   49    6505.155419        No  Travel_Frequently        279   
2   37    2090.000000       Yes      Travel_Rarely       1373   
3   33    2909.000000        No  Travel_Frequently       1392   
4   27    3468.000000        No      Travel_Rarely        591   

               Department  DistanceFromHome  Education EducationField  \
0                   Sales                 1          2  Life Sciences   
1  Research & Development                 8          1  Life Sciences   
2  Research & Development                 2          2          Other   
3  Research & Development                 3          4  Life Sciences   
4  Research & Development                 2          1        Medical   

   EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  HourlyRate  \
0              1             1.0                        2  Female          94   
1              1             2.0                        3    Male          61   
2              1             4.0                        4    Male          92   
3              1             5.0                        4  Female          56   
4              1             7.0                        1    Male          40   

   JobInvolvement  JobLevel                JobRole  JobSatisfaction  \
0               3         2        Sales Executive                4   
1               2         2     Research Scientist                2   
2               2         1  Laboratory Technician                3   
3               3         1     Research Scientist                3   
4               3         1  Laboratory Technician                2   

  MaritalStatus  MonthlyRate  NumCompaniesWorked Over18 OverTime  \
0        Single      19479.0                   8      Y      Yes   
1       Married      24907.0                   1      Y       No   
2        Single       2396.0                   6      Y      Yes   
3       Married      23159.0                   1      Y      Yes   
4       Married      16632.0                   9      Y       No   

   PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
0               11.0                3.0                         1   
1               23.0                4.0                         4   
2               15.0                3.0                         2   
3               11.0                3.0                         3   
4               12.0                3.0                         4   

   StandardHours  StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0             80                 0                  8                      0   
1             80                 1                 10                      3   
2             80                 0                  7                      3   
3             80                 0                  8                      3   
4             80                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [326]:
# filter employees under 20 age

mask_age_U20 = df['Age'] < 20
df[mask_age_U20].sample(5)

Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
1153   18         1569.0       Yes  Travel_Frequently        544   
853    19         2552.0        No      Travel_Rarely        645   
828    18         1904.0       Yes         Non-Travel        247   
177    19         1102.0       Yes      Travel_Rarely        303   
457    18         1878.0       Yes  Travel_Frequently       1306   

                  Department  DistanceFromHome  Education EducationField  \
1153                   Sales                 3          2        Medical   
853   Research & Development                 9          2  Life Sciences   
828   Research & Development                 8          1        Medical   
177   Research & Development                 2          3  Life Sciences   
457                    Sales                 5          3      Marketing   

      EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  \
1153              1          1624.0                        2  Female   
853               1          1193.0                        3    Male   
828               1          1156.0                        3    Male   
177               1           243.0                        2    Male   
457               1           614.0                        2    Male   

      HourlyRate  JobInvolvement  JobLevel                JobRole  \
1153          70               3         1   Sales Representative   
853           54               3         1     Research Scientist   
828           80               3         1  Laboratory Technician   
177           47               2         1  Laboratory Technician   
457           69               3         1   Sales Representative   

      JobSatisfaction MaritalStatus  MonthlyRate  NumCompaniesWorked Over18  \
1153                4        Single      18420.0                   1      Y   
853                 1        Single       7172.0                   1      Y   
828                 3        Single      13556.0                   1      Y   
177                 4        Single       9241.0                   1      Y   
457                 2        Single       8059.0                   1      Y   

     OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
1153      Yes               12.0                3.0                         3   
853        No               25.0                4.0                         3   
828        No               12.0                3.0                         4   
177        No               22.0                4.0                         3   
457       Yes               14.0                3.0                         4   

      StandardHours  StockOptionLevel  TotalWorkingYears  \
1153             80                 0                  0   
853              80                 0                  1   
828              80                 0                  0   
177              80                 0                  1   
457              80                 0                  0   

      TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
1153                      2                4               0   
853                       4                3               1   
828                       0                3               0   
177                       3                2               1   
457                       3                3               0   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
1153                   0                        0                     0  
853                    1                        0                     0  
828                    0                        0                     0  
177                    0                        1                     0  
457                    0                        0                     0

In [325]:
# filter employees travel_frequently AND married

mask_married_travel = (df['BusinessTravel'] == 'Travel_Frequently')  &  (df['MaritalStatus'] == 'Married')
df[mask_married_travel].sample(5)

Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
831   31         2610.0       Yes  Travel_Frequently        874   
548   43         6804.0        No  Travel_Frequently        775   
845   40         4422.0        No  Travel_Frequently        902   
762   26         2042.0       Yes  Travel_Frequently        342   
994   52        13247.0        No  Travel_Frequently        322   

                 Department  DistanceFromHome  Education EducationField  \
831  Research & Development                15          3        Medical   
548                   Sales                15          3  Life Sciences   
845  Research & Development                26          2        Medical   
762  Research & Development                 2          3  Life Sciences   
994  Research & Development                28          2        Medical   

     EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  \
831              1          1160.0                        3    Male   
548              1           754.0                        4    Male   
845              1          1180.0                        3  Female   
762              1          1053.0                        1    Male   
994              1          1401.0                        4  Female   

     HourlyRate  JobInvolvement  JobLevel                 JobRole  \
831          72               3         1   Laboratory Technician   
548          47               2         2         Sales Executive   
845          92               2         2      Research Scientist   
762          57               3         1      Research Scientist   
994          59               4         4  Manufacturing Director   

     JobSatisfaction MaritalStatus  MonthlyRate  NumCompaniesWorked Over18  \
831                3       Married       6233.0                   1      Y   
548                4       Married      23683.0                   3      Y   
845                4       Married      21203.0                   3      Y   
762                1       Married      15346.0                   6      Y   
994                3       Married       9731.0                   2      Y   

    OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
831       No               12.0                3.0                         3   
548       No               18.0                3.0                         3   
845      Yes               13.0                3.0                         4   
762      Yes               14.0                3.0                         2   
994      Yes               11.0                3.0                         2   

     StandardHours  StockOptionLevel  TotalWorkingYears  \
831             80                 1                  2   
548             80                 1                  7   
845             80                 1                 16   
762             80                 1                  6   
994             80                 1                 24   

     TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
831                      5                2               2   
548                      5                3               2   
845                      3                1               1   
762                      2                3               3   
994                      3                2               5   

     YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
831                   2                        2                     2  
548                   2                        2                     2  
845                   1                        0                     0  
762                   2                        1                     2  
994                   3                        0                     2

In [324]:
# filter employees have monthly income between 5000-7000 USD

mask_income_between = df['MonthlyIncome'].between(5000, 7000)
df[mask_income_between].sample(5)

Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
566   47         6397.0       Yes  Travel_Frequently        719   
973   35         5363.0        No      Travel_Rarely        817   
745   55         6385.0        No  Travel_Frequently        135   
353   37         5974.0        No      Travel_Rarely       1319   
378   34         5304.0       Yes         Non-Travel       1362   

                 Department  DistanceFromHome  Education EducationField  \
566                   Sales                27          2  Life Sciences   
973  Research & Development                 1          3        Medical   
745  Research & Development                18          4        Medical   
353  Research & Development                 6          3        Medical   
378                   Sales                19          3      Marketing   

     EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  \
566              1           785.0                        2  Female   
973              1          1369.0                        4  Female   
745              1          1034.0                        3    Male   
353              1           474.0                        3    Male   
378              1           502.0                        1    Male   

     HourlyRate  JobInvolvement  JobLevel                    JobRole  \
566          77               4         2            Sales Executive   
973          60               2         2      Laboratory Technician   
745          62               3         2  Healthcare Representative   
353          51               4         2         Research Scientist   
378          67               4         2            Sales Executive   

     JobSatisfaction MaritalStatus  MonthlyRate  NumCompaniesWorked Over18  \
566                3        Single      10339.0                   4      Y   
973                4       Married      10846.0                   0      Y   
745                2       Married      12992.0                   3      Y   
353                1      Divorced      17001.0                   4      Y   
378                4        Single       4652.0                   8      Y   

    OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
566      Yes               12.0                3.0                         4   
973       No               12.0                3.0                         2   
745      Yes               14.0                3.0                         4   
353      Yes               13.0                3.0                         1   
378      Yes               13.0                3.0                         2   

     StandardHours  StockOptionLevel  TotalWorkingYears  \
566             80                 0                  8   
973             80                 1                 10   
745             80                 2                 17   
353             80                 2                 13   
378             80                 0                  9   

     TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
566                      2                3               5   
973                      0                3               9   
745                      3                3               8   
353                      2                3               7   
378                      3                2               5   

     YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
566                   4                        1                     3  
973                   7                        0                     0  
745                   7                        6                     7  
353                   7                        6                     7  
378                   2                        0                     4

In [327]:
# Lets see high valued employees who has a DailyRate of 1450 and above

mask_daily_rate_high = df['DailyRate'] > 1450
df[mask_daily_rate_high].sample(5)

Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
1163   37         3936.0        No      Travel_Rarely       1470   
110    51         7484.0        No  Travel_Frequently       1456   
469    32         4707.0       Yes         Non-Travel       1474   
1044   49         6651.0        No      Travel_Rarely       1495   
412    46         4810.0        No      Travel_Rarely       1485   

                  Department  DistanceFromHome  Education    EducationField  \
1163  Research & Development                10          3           Medical   
110   Research & Development                 1          4           Medical   
469                    Sales                11          4             Other   
1044  Research & Development                 5          4  Technical Degree   
412   Research & Development                18          3           Medical   

      EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  \
1163              1          1640.0                        2  Female   
110               1           145.0                        1  Female   
469               1           631.0                        4    Male   
1044              1          1473.0                        1    Male   
412               1           550.0                        3  Female   

      HourlyRate  JobInvolvement  JobLevel                    JobRole  \
1163          71               3         1         Research Scientist   
110           30               2         3  Healthcare Representative   
469           60               4         2            Sales Executive   
1044          96               3         2  Healthcare Representative   
412           87               3         2     Manufacturing Director   

      JobSatisfaction MaritalStatus  MonthlyRate  NumCompaniesWorked Over18  \
1163                2       Married       9953.0                   1      Y   
110                 1        Single      25796.0                   3      Y   
469                 3       Married      23914.0                   8      Y   
1044                3       Married      21534.0                   2      Y   
412                 3      Divorced      26314.0                   2      Y   

     OverTime  PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
1163       No               11.0                3.0                         1   
110        No               20.0                4.0                         3   
469        No               12.0                3.0                         4   
1044       No               14.0                3.0                         2   
412        No               14.0                3.0                         3   

      StandardHours  StockOptionLevel  TotalWorkingYears  \
1163             80                 1                  8   
110              80                 0                 23   
469              80                 0                  6   
1044             80                 1                 20   
412              80                 1                 19   

      TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
1163                      2                1               8   
110                       1                2              13   
469                       2                3               4   
1044                      0                2               3   
412                       5                2              10   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
1163                   4                        7                     7  
110                   12                       12                     8  
469                    2                        1                     2  
1044                   2                        1                     2  
412                    7                        0                     8

In [329]:
# The manager is also asking 'the combined Daily Rate for all employees with 1450+"

df[mask_daily_rate_high]['DailyRate'].sum()

70849

In [337]:
print(f"{len(df[mask_daily_rate_high])} employees are making 1450+ Daily Rate\n"  
      f"within {len(df)} employees this total {len(df[mask_daily_rate_high])} employees are generating {df[mask_daily_rate_high]['DailyRate'].sum()} cumulatively" )

48 employees are making 1450+ Daily Rate
within 1470 employees this total 48 employees are generating 70849 cumulatively


# TASK-6: PERFORM BASIC EDA ON BOTH CLASSES

## Dropping some Columns

In [340]:
# dataset shape before dropping any values

df.shape

(1470, 35)

**It makes sense to drop 'EmployeeCount', 'Standardhours' and 'Over18' since they do not change from one employee to the other**

In [342]:
df.drop(columns=['EmployeeCount', 'StandardHours', 'Over18'], inplace=True)

In [343]:
# dataset shape AFTER dropping any values

df.shape

(1470, 32)

In [ ]:
# We can also drop the employee number aswell
# we will drop with some other parametres

df.drop('EmployeeNumber', axis=1, inplace=True)

In [349]:
# dataset shape AFTER dropping any values

df.shape

(1470, 31)

In [350]:
df.head()

Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
0   41    5993.000000       Yes      Travel_Rarely       1102   
1   49    6505.155419        No  Travel_Frequently        279   
2   37    2090.000000       Yes      Travel_Rarely       1373   
3   33    2909.000000        No  Travel_Frequently       1392   
4   27    3468.000000        No      Travel_Rarely        591   

               Department  DistanceFromHome  Education EducationField  \
0                   Sales                 1          2  Life Sciences   
1  Research & Development                 8          1  Life Sciences   
2  Research & Development                 2          2          Other   
3  Research & Development                 3          4  Life Sciences   
4  Research & Development                 2          1        Medical   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyRate  \
0        Sales Executive                4        Single      19479.0   
1     Research Scientist                2       Married      24907.0   
2  Laboratory Technician                3        Single       2396.0   
3     Research Scientist                3       Married      23159.0   
4  Laboratory Technician                2       Married      16632.0   

   NumCompaniesWorked OverTime  PercentSalaryHike  PerformanceRating  \
0                   8      Yes               11.0                3.0   
1                   1       No               23.0                4.0   
2                   6      Yes               15.0                3.0   
3                   1      Yes               11.0                3.0   
4                   9       No               12.0                3.0   

   RelationshipSatisfaction  StockOptionLevel  TotalWorkingYears  \
0                         1                 0                  8   
1                         4                 1                 10   
2                         2                 0                  7   
3                         3                 0                  8   
4                         4                 1                  6   

   TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                      0                1               6                   4   
1                      3                3              10                   7   
2                      3                3               0                   0   
3                      3                3               8                   7   
4                      3                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

## A comparison between two Classes

**Lets focus on ATTRITION column to gather some insight about the employees who left the company VS the one who stayed**

In [357]:
# lets filter the data with ones who stayed in the company

df[df['Attrition'] == 'Yes']

Age  MonthlyIncome Attrition     BusinessTravel  DailyRate  \
0      41    5993.000000       Yes      Travel_Rarely       1102   
2      37    2090.000000       Yes      Travel_Rarely       1373   
14     28    2028.000000       Yes      Travel_Rarely        103   
21     36    3407.000000       Yes      Travel_Rarely       1218   
24     34    2960.000000       Yes      Travel_Rarely        699   
26     32    3919.000000       Yes  Travel_Frequently       1125   
33     39    2086.000000       Yes      Travel_Rarely        895   
34     24    2293.000000       Yes      Travel_Rarely        813   
36     50    2683.000000       Yes      Travel_Rarely        869   
42     26    2293.000000       Yes      Travel_Rarely       1357   
45     41   19545.000000       Yes      Travel_Rarely       1360   
50     48    5381.000000       Yes      Travel_Rarely        626   
51     28    3441.000000       Yes      Travel_Rarely       1434   
69     36    3388.000000       Yes      Travel_Rarely        318   
89     46    9619.000000       Yes      Travel_Rarely        669   
100    37    2073.000000       Yes      Travel_Rarely        807   
102    20    2926.000000       Yes  Travel_Frequently        871   
107    25    5744.000000       Yes      Travel_Rarely        240   
111    34    6074.000000       Yes  Travel_Frequently        658   
122    56    4963.000000       Yes      Travel_Rarely        441   
124    31    6172.000000       Yes      Travel_Rarely        249   
126    58   10312.000000       Yes      Travel_Rarely        147   
127    19    1675.000000       Yes      Travel_Rarely        528   
132    31    4559.000000       Yes      Travel_Rarely        542   
136    51   10650.000000       Yes  Travel_Frequently       1150   
140    32    4200.000000       Yes      Travel_Rarely       1033   
171    19    2325.000000       Yes  Travel_Frequently        602   
177    19    1102.000000       Yes      Travel_Rarely        303   
182    41    3140.000000       Yes      Travel_Rarely       1356   
192    35    5916.000000       Yes      Travel_Rarely        556   
204    38    6673.000000       Yes      Travel_Rarely       1180   
205    29    7639.000000       Yes      Travel_Rarely        121   
210    32   10400.000000       Yes      Travel_Rarely       1045   
214    30    2657.000000       Yes      Travel_Rarely       1005   
216    30    6696.000000       Yes  Travel_Frequently        334   
217    29    2058.000000       Yes      Travel_Rarely        992   
229    29    2389.000000       Yes      Travel_Rarely        896   
234    33    2436.000000       Yes      Travel_Rarely        813   
236    33    2707.000000       Yes      Travel_Rarely        465   
239    32    3730.000000       Yes      Travel_Rarely        515   
250    37   10048.000000       Yes  Travel_Frequently        504   
259    31    3479.000000       Yes  Travel_Frequently        307   
264    28    3485.000000       Yes      Travel_Rarely        529   
271    47   11849.000000       Yes         Non-Travel        666   
286    44    3161.000000       Yes  Travel_Frequently        920   
288    26    2373.000000       Yes      Travel_Rarely       1449   
293    26    5828.000000       Yes      Travel_Rarely        950   
296    18    1420.000000       Yes      Travel_Rarely        230   
317    52    4941.000000       Yes      Travel_Rarely        723   
323    28    3464.000000       Yes      Travel_Rarely       1157   
327    39    5238.000000       Yes      Travel_Rarely       1162   
336    29    2119.000000       Yes      Travel_Rarely        318   
357    21    2174.000000       Yes  Travel_Frequently        756   
363    33    2851.000000       Yes      Travel_Rarely        350   
366    41    9355.000000       Yes  Travel_Frequently        143   
368    40    6380.000000       Yes      Travel_Rarely        575   
370    21    2716.000000       Yes      Travel_Rarely        156   
378    34    5304.000000       Yes         Non-Travel       13

In [359]:
#lets see some statistics of employees stayed in the company

df[df['Attrition'] == 'Yes'].describe().T

count          mean          std     min     25%  \
Age                       237.0     33.607595     9.689350    18.0    28.0   
MonthlyIncome             237.0   4778.903609  3634.122234  1009.0  2373.0   
DailyRate                 237.0    750.362869   401.899519   103.0   408.0   
DistanceFromHome          237.0     10.632911     8.452525     1.0     3.0   
Education                 237.0      2.839662     1.008244     1.0     2.0   
EnvironmentSatisfaction   237.0      2.464135     1.169791     1.0     1.0   
HourlyRate                237.0     65.573840    20.099958    31.0    50.0   
JobInvolvement            237.0      2.518987     0.773405     1.0     2.0   
JobLevel                  237.0      1.637131     0.940594     1.0     1.0   
JobSatisfaction           237.0      2.468354     1.118058     1.0     1.0   
MonthlyRate               237.0  14559.308017  7208.153264  2326.0  8870.0   
NumCompaniesWorked        237.0      2.940928     2.678519     0.0     1.0   
PercentSalaryHike         237.0     15.097046     3.770294    11.0    12.0   
PerformanceRating         237.0      3.156118     0.363735     3.0     3.0   
RelationshipSatisfaction  237.0      2.599156     1.125437     1.0     2.0   
StockOptionLevel          237.0      0.527426     0.856361     0.0     0.0   
TotalWorkingYears         237.0      8.244726     7.169204     0.0     3.0   
TrainingTimesLastYear     237.0      2.624473     1.254784     0.0     2.0   
WorkLifeBalance           237.0      2.658228     0.816453     1.0     2.0   
YearsAtCompany            237.0      5.130802     5.949984     0.0     1.0   
YearsInCurrentRole        237.0      2.902954     3.174827     0.0     0.0   
YearsSinceLastPromotion   237.0      1.945148     3.153077     0.0     0.0   
YearsWithCurrManager      237.0      2.852321     3.143349     0.0     0.0   

                              50%      75%      max  
Age                          32.0     39.0     58.0  
MonthlyIncome              3202.0   5916.0  19859.0  
DailyRate                   699.0   1092.0   1496.0  
DistanceFromHome              9.0     17.0     29.0  
Education                     3.0      4.0      5.0  
EnvironmentSatisfaction       3.0      4.0      4.0  
HourlyRate                   66.0     84.0    100.0  
JobInvolvement                3.0      3.0      4.0  
JobLevel                      1.0      2.0      5.0  
JobSatisfaction               3.0      3.0      4.0  
MonthlyRate               14618.0  21081.0  26999.0  
NumCompaniesWorked            1.0      5.0      9.0  
PercentSalaryHike            14.0     17.0     25.0  
PerformanceRating             3.0      3.0      4.0  
RelationshipSatisfaction      3.0      4.0      4.0  
StockOptionLevel              0.0      1.0      3.0  
TotalWorkingYears             7.0     10.0     40.0  
TrainingTimesLastYear         2.0      3.0      6.0  
WorkLifeBalance               3.0      3.0      4.0  
YearsAtCompany                3.0      7.0     40.0  
YearsInCurrentRole            2.0      4.0     15.0  
YearsSinceLastPromotion       1.0      2.0     15.0  
YearsWithCurrManager          2.0      5.0     14.0

In [370]:
df_comparison = pd.DataFrame(columns=['Count_Stayed', 'Count_Left', 'Mean_Stayed', 'Mean_Left'], index=df[df['Attrition'] == 'Yes'].describe().T.index)
df_comparison['Count_Stayed'] = df[df['Attrition'] == 'Yes'].count()
df_comparison['Count_Left'] = df[df['Attrition'] == 'No'].count()
df_comparison['Mean_Stayed'] = df[df['Attrition'] == 'Yes'].mean()
df_comparison['Mean_Left'] = df[df['Attrition'] == 'No'].mean()
df_comparison

Count_Stayed  Count_Left   Mean_Stayed     Mean_Left
Age                                237        1233     33.607595     37.561233
MonthlyIncome                      237        1233   4778.903609   6836.965378
DailyRate                          237        1233    750.362869    812.504461
DistanceFromHome                   237        1233     10.632911      8.915653
Education                          237        1233      2.839662      2.927007
EnvironmentSatisfaction            237        1233      2.464135      2.771290
HourlyRate                         237        1233     65.573840     65.952149
JobInvolvement                     237        1233      2.518987      2.770479
JobLevel                           237        1233      1.637131      2.145985
JobSatisfaction                    237        1233      2.468354      2.778589
MonthlyRate                        237        1233  14559.308017  14263.231900
NumCompaniesWorked                 237        1233      2.940928      2.645580
PercentSalaryHike                  237        1233     15.097046     15.227256
PerformanceRating                  237        1233      3.156118      3.152598
RelationshipSatisfaction           237        1233      2.599156      2.733982
StockOptionLevel                   237        1233      0.527426      0.845093
TotalWorkingYears                  237        1233      8.244726     11.862936
TrainingTimesLastYear              237        1233      2.624473      2.832928
WorkLifeBalance                    237        1233      2.658228      2.781022
YearsAtCompany                     237        1233      5.130802      7.369019
YearsInCurrentRole                 237        1233      2.902954      4.484185
YearsSinceLastPromotion            237        1233      1.945148      2.234388
YearsWithCurrManager               237        1233      2.852321      4.367397

- 1233 employees have left the company whereas 237 are still working in the company.
- The monthly income of employees who left the company is significantly higher than the ones who stayed in
- Also the mean of Age is higher for the employees who left the company.
- Moreover, jobInvolvement is also far much higher amongt the left employees
- Also, normally stock option level is also higher for the employees who left the company due to their higher monthly income, higher job Level and jobInvolvementow()

# TASK-7: PANDAS WITH FUNCTIONS

Let's assume the daily rate has increased by 10%
Define a function that increases all clients networth by a fixed value of 10% (for simplicity sake)
assume that net worth increased by 10%
You can apply a function to the DataFrame